tests/test-helmet.php

```diff
@@ -0,0 +1,317 @@
+			[
+				$meta_example,
+				$default_title,
+			],
+			$data_with_helmet['defaults'][0]->get_children()
+		);
+
+		/**
+		 * Test the page helmet filter.
+		 */
+		add_filter(
+			'wp_irving_page_helmet_component',
+			function( $helmet ) use ( $meta_example ) {
+				$helmet->prepend_child( $meta_example ); // Pre-pend so we can test more complex output.
+				return $helmet;
+			}
+		);
+
+		// Run setup with the new helmet filter in place.
+		$data_with_helmet = Templates\setup_helmet( $data, new \WP_Query(), 'site', '/', new \WP_Rest_Request() );
+		$this->assertEquals(
+			[
+				$meta_example,
+				$page_title,
+			],
+			$data_with_helmet['page'][0]->get_children()
+		);
+	}
+
+	/**
+	 * Test the `inject_head_tags` function.
+	 */
+	public function test_inject_head_tags() {
+
+		// Example components.
+		$helmet = new Component( 'irving/helmet' );
+		$title  = ( new Component( 'title' ) )->set_child( 'Irving Development - Just another WordPress site' );
+
+		// Parse the example markup, only grabbing the title tag.
+		$parsed_html = parse_html( $this->example_markup, [ 'title' ] );
+
+		$this->assertEquals(
+			( new Component( 'irving/helmet' ) )->set_child( $title ),
+			inject_head_tags( $helmet, $parsed_html )
+		);
+	}
+
+	/**
+	 * Test the `get_title_from_helmet()` functin.
+	 */
+	public function test_get_title_from_helmet() {
+
+		// Helmet has no children.
+		$helmet = new Component( 'irving/helmet' );
+		$this->assertNull( get_title_from_helmet( $helmet ) );
+
+		// Helmet->Title has no children.
+		$helmet->set_child( new Component( 'title' ) );
+		$this->assertNull( get_title_from_helmet( $helmet ) );
+
+		// Helmet->Title->first child is a component.
+		$helmet->set_child( ( new Component( 'title' ) )->set_child( new Component( 'irving/text' ) ) );
+		$this->assertNull( get_title_from_helmet( $helmet ) );
+
+		// Title is `Foo Bar`.
+		$helmet->set_child( ( new Component( 'title' ) )->set_child( 'Foo Bar' ) );
+		$this->assertEquals( 'Foo Bar', get_title_from_helmet( $helmet ) );
+	}
+
+
+	/**
+	 * Helper to get a Helmet component without a title.
+	 *
+	 * @return Component
+	 */
+	public function get_no_title() {
+		return new Component( 'irving/helmet' );
+	}
+
+	/**
+	 * Helper to get a Helmet component with a title of `Foo Bar`.
+	 *
+	 * @return Component
+	 */
+	public function get_foo_bar_title() {
+		return ( new Component( 'irving/helmet' ) )
+			->prepend_child(
+				new Component(
+					'title',
+					[
+						'children' => [ 'Foo Bar' ]
+					]
+				)
+			);
+	}
+
+	/**
+	 * Helper to get a Helmet component with a title of `Hello World`.
+	 *
+	 * @return Component
+	 */
+	public function get_hello_world_title() {
+		return ( new Component( 'irving/helmet' ) )
+			->prepend_child(
+				new Component(
+					'title',
+					[
+						'children' => [ 'Hello World' ]
+					]
+				)
+			);
+	}
+
+	/**
+	 * Test the `create_or_update_title` function.
+	 */
+	public function test_create_or_update_title() {
```

Comment:
This could also be broken up into a data provider or perhaps be broken into distinct methods to test creation and updating paths.